pkg/admissioncontroller/webhooks/admission/auditpolicy/admission.go

```diff
@@ -0,0 +1,285 @@
+
+	shootGVK     = metav1.GroupVersionKind{Group: "core.gardener.cloud", Kind: "Shoot", Version: "v1beta1"}
+	configmapGVK = metav1.GroupVersionKind{Group: "", Kind: "ConfigMap", Version: "v1"}
+)
+
+// New creates a new handler for validating audit policies.
+func New(logger logr.Logger) *handler {
+	return &handler{
+		logger: logger,
+	}
+}
+
+type handler struct {
+	apiReader client.Reader
+	decoder   *admission.Decoder
+	logger    logr.Logger
+}
+
+var _ admission.Handler = &handler{}
+
+func (h *handler) InjectAPIReader(reader client.Reader) error {
+	h.apiReader = reader
+	return nil
+}
+
+func (h *handler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+// Handle implements the webhook handler for audit log policy validation
+func (h *handler) Handle(ctx context.Context, request admission.Request) admission.Response {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	switch request.Kind {
+	case shootGVK:
+		return h.admitShoot(ctx, request)
+	case configmapGVK:
+		return h.admitConfigMap(ctx, request)
+	}
+	return acadmission.Allowed("resource is not core.gardener.cloud/v1beta1.shoot or v1.configmap")
+}
+
+func (h *handler) admitShoot(ctx context.Context, request admission.Request) admission.Response {
+	var (
+		shoot    = &gardencorev1beta1.Shoot{}
+		oldShoot = &gardencorev1beta1.Shoot{}
+	)
+	if request.Operation != admissionv1.Create && request.Operation != admissionv1.Update {
+		return acadmission.Allowed("operation is not Create or Update")
+	}
+
+	if err := h.decoder.Decode(request, shoot); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if !hasAuditPolicy(shoot) {
+		return acadmission.Allowed("shoot resource is not specifying any audit policy")
+	}
+	cmRef := shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.ConfigMapRef
+
+	if request.Operation == admissionv1.Update {
+		if err := h.getOldObject(request, oldShoot); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+
+		if hasAuditPolicy(oldShoot) && reflect.DeepEqual(cmRef, oldShoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.ConfigMapRef) {
+			return acadmission.Allowed("audit policy reference did not change in Shoot resource")
+		}
+	}
+
+	auditPolicyCm := &corev1.ConfigMap{}
+	if err := h.apiReader.Get(ctx, kutil.Key(shoot.Namespace, cmRef.Name), auditPolicyCm); err != nil {
+		if apierrors.IsNotFound(err) {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("referenced audit policy does not exist: namespace: %s, name: %s", cmRef.Namespace, cmRef.Name))
+		}
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("could not retrieve config map: %s", err))
+	}
+
+	auditPolicy, err := getAuditPolicy(auditPolicyCm)
+	if err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("Error: %s in %v/%v", err, cmRef.Namespace, cmRef.Name))
```

Comment:
`%v` / `%w` should be used for formatting errors.
Nit: usually the error is also appended and not prepended, otherwise further wrapping will look weird.
```suggestion
		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("error getting auditlog policy from ConfigMap %s/%s: %w", cmRef.Namespace, cmRef.Name, err))
```

Reply:
👍🏼